File: sdks/typescript/examples/quickstart.ts
/**
 * CostObs TypeScript SDK quickstart — runnable offline against the local stack.
 *
 * Run:  npx tsx examples/quickstart.ts
 *
 * If COSTOBS_API_KEY / a real OpenAI client are not configured, this uses a
 * fake provider client so the full loop (wrap -> call -> cost calc -> enqueue
 * -> ship) runs end to end without any network calls leaving your machine.
 *
 * Against the local stack, point it at the ingest service with the dev key:
 *   COSTOBS_INGEST_URL=http://localhost:8080 \
 *   COSTOBS_API_KEY=costobs_dev_secret_key \
 *   npx tsx examples/quickstart.ts
 */

import { configure, flush, requestContext, shutdown, wrap } from "../src/index.js";

// ---- A fake OpenAI-shaped client (constructor name drives detection) ----
class OpenAI {
  chat = {
    completions: {
      create: (_opts: Record<string, unknown>) => ({
        id: "chatcmpl-demo",
        model: "gpt-4o-2024-08-06",
        usage: {
          prompt_tokens: 1200,
          completion_tokens: 300,
          total_tokens: 1500,
          prompt_tokens_details: { cached_tokens: 800 },
          completion_tokens_details: { reasoning_tokens: 0 },
        },
      }),
    },
  };
  embeddings = { create: (_opts: Record<string, unknown>) => ({ usage: {} }) };
}

async function main(): Promise<void> {
  // Dev defaults for the local stack. Env vars override these.
  configure({
    ingestUrl: process.env.COSTOBS_INGEST_URL ?? "http://localhost:8080",
    apiKey: process.env.COSTOBS_API_KEY ?? "costobs_dev_secret_key",
    flushIntervalMs: 250,
  });

  // 1. Wrap the provider client once. Client-level tags apply to every event.
  const client = wrap(new OpenAI(), {
    team: "growth",
    environment: "prod",
    service: "quickstart",
  });

  // 2. A plain call. Cost is computed locally and shipped in the background;
  //    the provider call always returns first.
  const resp = client.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: "Say hi" }],
    // Call-site metadata: stripped before the provider ever sees it.
    costobs: { feature: "greeting", promptVersion: "v3" },
  });
  console.log("provider response id:", (resp as { id: string }).id);

  // 3. requestContext attaches per-request metadata to every nested LLM call,
  //    propagating across awaits (AsyncLocalStorage).
  await requestContext(
    { customerId: "cust_42", traceId: "trace_abc" },
    async () => {
      client.chat.completions.create({
        model: "gpt-4o",
        messages: [{ role: "user", content: "And again" }],
        costobs: { feature: "followup" },
      });
      await Promise.resolve();
    },
  );

  // 4. Drain telemetry before exit (flush) and stop background timers (shutdown).
  await flush(2000);
  await shutdown(2000);
  console.log("done — sent 2 events to", process.env.COSTOBS_INGEST_URL ?? "http://localhost:8080");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: sdks/python/examples/quickstart.py
"""Runnable CostObs SDK demo.

Sends events through the full SDK -> ingest loop. If no real OPENAI_API_KEY is
present, it uses a fake OpenAI-shaped client so the demo runs fully offline and
still exercises wrap -> cost calc -> background ship.

Run against the local stack::

    COSTOBS_INGEST_URL=http://localhost:8080 \
    COSTOBS_API_KEY=costobs_dev_secret_key \
    python examples/quickstart.py
"""

from __future__ import annotations

import logging
import os
import sys
import types

logging.basicConfig(level=logging.INFO)

import costobs


def _fake_openai_client():
    """A minimal OpenAI-shaped client (module name 'openai' drives detection)."""
    if "openai" not in sys.modules:
        sys.modules["openai"] = types.ModuleType("openai")

    class _Usage:
        prompt_tokens = 1200
        completion_tokens = 350
        total_tokens = 1550
        prompt_tokens_details = type("D", (), {"cached_tokens": 800})()
        completion_tokens_details = type("D", (), {"reasoning_tokens": 0})()

    class _Resp:
        usage = _Usage()
        model = "gpt-4o"

    class _Completions:
        def create(self, **kwargs):
            return _Resp()

    class _Chat:
        def __init__(self):
            self.completions = _Completions()

    class OpenAI:
        __module__ = "openai"

        def __init__(self):
            self.chat = _Chat()

    OpenAI.__module__ = "openai"
    return OpenAI()


def main() -> None:
    costobs.configure(
        ingest_url=os.environ.get("COSTOBS_INGEST_URL", "http://localhost:8080"),
        api_key=os.environ.get("COSTOBS_API_KEY", "costobs_dev_secret_key"),
    )

    if os.environ.get("OPENAI_API_KEY"):
        from openai import OpenAI

        raw = OpenAI()
        print("Using real OpenAI client.")
    else:
        raw = _fake_openai_client()
        print("No OPENAI_API_KEY set — using offline fake client.")

    client = costobs.wrap(raw, team="demo", environment="dev", service="quickstart")

    with costobs.request_context(customer_id="cust-42", trace_id="trace-abc"):
        resp = client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "user", "content": "Summarize CostObs in one line."}],
            feature="summarize",
            prompt_version="v1",
        )
        print("Provider response model:", getattr(resp, "model", "?"))

    # Drain telemetry so the demo doesn't exit before the event ships.
    costobs.flush(timeout=5.0)
    costobs.shutdown(timeout=5.0)
    print("Done. Event shipped (or logged a warning if ingest was unreachable).")


if __name__ == "__main__":
    main()
